File: USACO/AlphaStar/Silver Diagnostic Exam/bcount.py
N, Q = map(int, input().split(" "))
cowBreeds = [int(input()) for i in range(N)]
queries = [list(map(int, input().split(" "))) for i in range(Q)]
for i in range(len(queries)):
    queries[i].append(i)

queries.sort(key=lambda x: x[0])

# print(N, Q)
# print(cowBreeds)
# print(queries)

# for query in queries:
#     a, b = query[0]-1, query[1]
#     breedCount = [0] * 3 # [Holsteins, Guernseys, Jerseys]
#     for i in range(a, b):
#         breedCount[cowBreeds[i]-1] += 1
#     print(*breedCount)

prev = [1, N]
breedCount = [0] * 3
answers = []
for i in cowBreeds:
    breedCount[i-1] += 1
for query in range(len(queries)):
    # Modify lower bound
    for i in range(prev[0]-1, queries[query][0]-1):
        breedCount[cowBreeds[i]-1] -= 1
    # Modify upper bound
    if prev[1] < queries[query][1]:
        for i in range(prev[1], queries[query][1]):
            breedCount[cowBreeds[i]-1] += 1
    elif prev[1] > queries[query][1]:
        for i in range(prev[1]-1, queries[query][1]-1, -1):
            breedCount[cowBreeds[i]-1] -= 1
    answers.append([*breedCount, queries[query][2]])
    # print(*breedCount)
    prev = queries[query]

answers.sort(key=lambda x: x[3])
for i in answers:
    print(*i[:3])

# [2, 1, 1, 3, 2, 1]
# [1, 6] to [2, 4]

File: USACO/AlphaStar/Silver Diagnostic Exam/bcount.java
import java.util.Scanner;
import java.util.Arrays;

public class bcount {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int N = input.nextInt();
        int Q = input.nextInt();
        int[] cowBreeds = new int[N];
        for (int i = 0; i < N; i++) {
            cowBreeds[i] = input.nextInt();
        }
        int[][] queries = new int[Q][2];
        for (int i = 0; i < Q; i++) {
            queries[i][0] = input.nextInt();
            queries[i][1] = input.nextInt();
        }
        input.close();
        int[] prev = {1, N};
        int[] breedCount = new int[3];
        int[][] answers = new int[Q][3];
        int a = 0;
        for (int i: cowBreeds) {
            breedCount[i-1] += 1;
        }
        for (int[] query: queries) {
            // Modify lower bound
            if (prev[0] < query[0]) {
                for (int i = prev[0]-1; i < query[0]-1; i++) {
                    breedCount[cowBreeds[i]-1] -= 1;
                }
            } else {
                for (int i = prev[0]-2; i > query[0]-2; i--) {
                    breedCount[cowBreeds[i]-1] += 1;
                }
            }
            // Modify upper bound
            if (prev[1] < query[1]) {
                for (int i = prev[1]; i < query[1]; i++) {
                    breedCount[cowBreeds[i]-1] += 1;
                }
            } else {
                for (int i = prev[1]-1; i > query[1]-1; i--) {
                    breedCount[cowBreeds[i]-1] -= 1;
                }
            }
            answers[a++] = new int[]{breedCount[0], breedCount[1], breedCount[2]};
            prev = query;
        }
        for (int[] i: answers) {
            System.out.printf("%s %s %s\n", i[0], i[1], i[2]);
        }
    }
}
